### DISCUSSION TITLE
Optimizers parameter space probing and not finding solution?

### DISCUSSION THREAD

**Discussion by lancelot49** [phoebe-project/phoebe2#809] [Questions]
Hello Kyle,

I am trying to fit and optimize a RV data, I know the period and the eccentricity of the system, however the rv geometry, pushes the eccentricity much higher, is there any way to run the estimator with a fixed eccentricity? Not my main problem however, I manually guessed my way  to a okay fit (picture below), but then trying to optimize with nelder for asini, q, vgamma and t0, never results in a solution, let it run for close to three hours yesterday but it never converged. How does nelder mead probe the parameter space? And is there any reason why it does not find a solution within few thousand iterations? Also, from what i see, when optimizing the rv, the tutorials online never include p0, even tough, changes to p0 slightly affect the rv?

Any help will be greatly appreciated,
cheers!
![Screenshot 2023-12-12 071918](https://github.com/phoebe-project/phoebe2/assets/99822226/a7f2b71a-4fa4-4b66-af62-cc190ea409ca)

**Comment by aprsa** [comment]
Not Kyle, but let me chime in anyway. :) What do you mean exactly by "never converges?" Do the parameter values not change, or do they get stuck at a local solution, do they push against the parameter limits, or something else entirely? If, for example, your asini wants to converge to negative values, that is the sign that primary and secondary RVs might be switched.

**Reply by lancelot49** [reply]
Hey, currently I run the nelder mead optimizer and the optimizer runs trough all the iterations I set it to, returning the original values, stating no solution found when I print out the results. As far as I know, there are no limits to the parameters but I have no idea what the parameters are doing during the optimization.

**Reply by aprsa** [reply]
NMS contracts the simplex around the minimum; if the parameters do not change, that means that no contraction results in an improvement of the cost function. That certainly indicates that something is fishy. Can you share your script/data?

**Reply by lancelot49** [reply]
Sure I can share

I create the system and since we are currently only interested in the rv part I disable the lc dataset

```
b = phoebe.default_binary() 
P1=23.943
b.set_value('latex_repr', component='binary', value='orb')
b.set_value('latex_repr', component='primary', value='1')
b.set_value('latex_repr', component='secondary', value='2')
b.set_value('period', component='binary', value=P1)
b.set_value('teff', component='primary', value=38000)
b.add_dataset('lc', 
              times=time_clean, 
              fluxes=fluxes_clean,      
              dataset='lc01')
b.add_dataset('rv', 
              times=time_rv, 
              rvs={'primary': rvs1, 'secondary': rvs2}, 
             sigmas={'primary': 0.05*np.abs(rvs1), 'secondary': 0.05*np.abs(rvs2)},
              dataset='rv01')
afig, mplfig = b.plot(x='phases',s=0.005, show=True)
b.disable_dataset('lc01')
```

Since I am manually setting the value of all the needed parameters, using rv geometry solution is not nessecary

```
b.flip_constraint('asini@binary', solve_for='sma@binary')
#b.adopt_solution('rv_geom_sol')
b.set_value_all('ecc@binary',0.3170040429616946)
b.set_value_all('t0_supconj',-4)
b.set_value_all('vgamma',-15) 
b.set_value_all('q',0.45)
b.set_value_all('asini@binary',120)
b.set_value_all('per0',85)
```

I know that q and v_gamma are interchangeable, they roughly follow the equation

```
b.set_value_all('pblum_mode', 'dataset-scaled')
b.set_value_all('atm', 'blackbody')
b.set_value_all('ld_mode', 'manual')
b.set_value_all('ld_mode_bol', 'manual')
b.run_compute(irrad_method='none', model='after_estimators', overwrite=True)
_ = b.plot(kind='rv',x='phases', m='.',s=0.005, show=True)
```

I have also tried with ellc fastcompute and phoebe compute but no change in result. Also ran it for roughly 3000 iterations yesterday, with no results
```
b.add_compute(compute='nm_fit',
              irrad_method='none',
              rv_method='dynamical',
              distortion_method='sphere')
b.add_solver('optimizer.nelder_mead',
             compute='nm_fit',overwrite=True,solver="nelder",adaptive=True)
b.set_value('maxiter', 1000)
b['fit_parameters'] = ['asini@binary','vgamma@system','q@binary','t0_supconj@binary']
```

Usually takes atleast half an hour to run for me, but in any case, returns the value set above.
Howeverm, might be of intrested, that usually, early on I get a warning:
RuntimeWarning: invalid value encountered in subtract
  np.max(np.abs(fsim[0] - fsim[1:])) <= fatol):
but no idea if that is a problem since it is able to continue running.
`b.run_solver("nelder", solution='nm_solution', overwrite=True)
print(b['nm_solution'])`


My data for the rv is
**rvs1**=[-75.01815849 -77.35339469 -85.96888407  33.88223277 -84.21097465
  20.59463503  22.76320734  13.75912754  34.29321021  53.17292247]

**rvs2**= [ 141.04953757  142.96513126  155.74338412 -110.89096184  141.85411335
 -106.49438885  -90.08026801  -10.1708171  -110.51669484 -128.57231569]

**time**=[3133.5887 3041.714  3135.5308 3363.796  3134.4997 3131.5603 5326.6924
 5324.5987 5327.5732 5328.4745]  in HDJ

**Reply by kecnry** [reply]
Do you have any priors enabled?  Can you enable `expose_lnprobabilities` in the optimizer run and make sure that the initial lnprobability is finite?  Does the `message` parameter in the solution say anything useful (that is passed on directly from the optimizer itself)?

**Reply by lancelot49** [reply]
Hey Kyle, I checked the expose_lnprobabilities and turns out worryingly that they are -inf, what does that mean and is there anyway to deal with it? As far as I can see when looking into the api doc that it only takes False or True so no idea if I can set it to some other value?

> ParameterSet: 13 parameters
R  fitted_twigs@nm_solution@so...: ['asini@binary@orbit@component' 'vgamma@system' 'q@binary@orbit@component'
 't0_supconj@binary@orbit@component']
   adopt_parameters@nm_solutio...: ['asini@binary@orbit@component', 'vgamma@system', 'q@binary@orbit@component', 't0_supconj@binary@orbit@component']
   adopt_distributions@nm_solu...: False
   adopt_values@nm_solution@so...: True
R    message@nm_solution@solution: Maximum number of iterations has been exceeded.
R      niter@nm_solution@solution: 1000
R    success@nm_solution@solution: False
R  initial_values@nm_solution@...: [120.0 -15.0 0.45 -4.0]
R  fitted_values@nm_solution@s...: [120.0 -15.0 0.45 -4.0]
R  fitted_units@nm_solution@so...: ['solRad' 'km / s' '' 'd']
R  initial_lnprobability@nm_so...: -inf
R  fitted_lnprobability@nm_sol...: -inf
    comments@nm_solution@solution:

**Reply by kecnry** [reply]
Ok, then that's the problem.  This has come up quite a few times lately, so I think a check that raises an error or warning when they start non-finite would be a good addition - I created #810 to eventually get this in.

A few ideas that might help to debug: Are you providing uncertainties on the RVs and are they non-zero?  Can you try with dynamical RVs (`rv_method` parameter) to see if perhaps this is coming from a phase where a component is completely eclipsed due to Rossiter-McLaughlin?  Try running just the forward model with `run_compute` and plotting or printing the residuals and make sure those make sense.

**Reply by lancelot49** [reply]
Hey Kyle,

All the data points have errors that are >0 and I was already running it with the dynamical rv method. Looking at the residuals
for the primary
>[−19.454205, 14.18295, 7.4911682, 5.8877022, −3.856874, −7.3969243, −0.010497372, 16.414273, 0.84851672, 9.0637206] 
for the secondary
> [65.907051, −12.115624, −3.6123798, −0.34765996, 11.622791, 4.0420223, 8.861911, 32.262049, 12.138276, 17.781563]
, other then the first point in the secondary being high, as we expected from the basic fit in the picture, they all seem reasonable.
Thank you so much for the help so far, but any other ideas that might be resulting in the probability being -inf?
​

**Reply by kecnry** [reply]
Nothing sticks out as being the culprit.... can you save your bundle immediately before calling `run_solver` (via `b.save('before_run_solver.bundle')`) and attach it here (note that that will also include your data - so if you're not okay with those being public, you can email the file instead)?

**Reply by lancelot49** [reply]
Had to zip it since github does not support the .bundle type, this was saved, as you said, just before I ran run_solver.
[before_run_solver.zip](https://github.com/phoebe-project/phoebe2/files/13652281/before_run_solver.zip)

**Reply by aprsa** [reply]
I managed to reproduce this and I didn't see anything glaringly obvious that would cause -infs. Please stay tuned, we're looking into this.

**Reply by kecnry** [reply]
Thanks for attaching the bundle, that helps a lot in tracking down what is happening.  Your solver is using the `nm_fit` compute options which fails when running `b.run_compute(compute='nm_fit')` with the following error:

```
ValueError: Could not compute intensities with atm='ck2004'.  Try changing atm to a table that covers a sufficient range of values (or to 'blackbody' in which case ld_mode will need to be set to 'manual' and coefficients provided via ld_coeffs).  Enable 'warning' logger to see out-of-bounds arrays.
```

Your `phoebe01` compute options use `atm='blackbody'`, which do succeed, which is probably why this was easily missed.  If you set the atmospheres to blackbody in `compute='nm_fit'`, then they do have finite likelihoods and the optimizer does run.  

With the light curve disabled and the RVs dynamical, this probably should not need to be called and can be fixed to avoid the error (I created #812 to track that problem).  For now, so long as you're only fitting the RVs, you can get around that by setting the atmospheres to blackbody.  But once you want to fit the light curves as well, I'd suggest re-enabling the atmospheres and ensuring that you're in a physical region of parameter space.  Currently your stars both have radii of 1 solRad, but the primary has a mass of 28 solMass and a temperature of 38000 K, which falls outside the bounds of the atmosphere table.

**Reply by lancelot49** [reply]
Thank you so much, just tried a few iteration run at it worked, would have never figured this out with your guys help!

**Reply by kecnry** [reply]
Just to follow-up, the bug with dynamical RVs failing with an atmosphere-out-of-bounds error has been fixed in #823 and will be included in the next (2.4.13) bugfix release.
